arachnado/rpc/ws.py

```diff
@@ -1,78 +1,67 @@
+    """
 
     def on_message(self, message):
-        self._results = []
         try:
-            msg = json.loads(message.encode('utf-8'))
+            msg = json.loads(message)
             event, data = msg['event'], msg['data']
         except (TypeError, ValueError):
-            logger.warn("Invalid message skipped" + message[:500])
+            logger.warn('Invalid message skipped: {!r}'.format(message[:500]))
             return
         if event == 'rpc:request':
-            self.on_event(data)
+            self.handle_request(json.dumps(data))
         else:
-            logger.warn("Unsupported event type: " + event)
-
-    def on_event(self, data):
-        self._RPC_.run(self, json_encode(data))
+            logger.warn('Unsupported event type: {!r}'.format(event))
 
-    def _result(self, result):
-        """A little hacky way to not close WS stream"""
-        self._RPC_finished = False
-        super(JsonRpcWebsocketHandler, self)._result(result)
+    def send_data(self, data):
+        self.write_event('rpc:response', data)
 
-    def on_result(self, data):
-        return self.write_event('rpc:response', data)
+    @gen.coroutine
+    def write_event(self, event, data):
+        if isinstance(data, six.string_types):
+            data = json.loads(data)
+        message = json_encode({'event': event, 'data': data})
+        try:
+            self.write_message(message)
+        except websocket.WebSocketClosedError:
+            pass
 
     def open(self):
-        """Forward open event to resource objects"""
-        for resource_name, resource in self.__dict__.iteritems():
-            if hasattr(RequestHandler, resource_name):
-                continue
+        """ Forward open event to resource objects.
+        """
+        for resource in self._resources():
             if hasattr(resource, '_on_open'):
                 resource._on_open()
-
-        self.__pinger = tornado.ioloop.PeriodicCallback(
-            lambda: self.ping('PING'),
-            1000 * 15
-        )
-        self.__pinger.start()
+        self._pinger = PeriodicCallback(lambda: self.ping(b'PING'), 1000 * 15)
+        self._pinger.start()
 
     def on_close(self):
-        """Forward on_close event to resource objects"""
-        self._RPC_finished = True
-        for resource_name, resource in self.__dict__.iteritems():
-            if hasattr(RequestHandler, resource_name):
-                continue
+        """ Forward on_close event to resource objects.
+        """
+        for resource in self._resources():
             if hasattr(resource, '_on_close'):
                 resource._on_close()
+        self._pinger.stop()
 
-        self.__pinger.stop()
-
-    @gen.coroutine
-    def write_event(self, event, data):
-        if isinstance(data, basestring):
-            data = json.loads(data)
-        message = json_encode({'event': event, 'data': data})
-        try:
-            msg_d = self.write_message(message)
```

Comment:
This is the thing I did not get - I can not see how this feature is used, and it looks like a Tornado 4.3-only feature: http://www.tornadoweb.org/en/stable/websocket.html#tornado.websocket.WebSocketHandler.write_message, so I just write the message in new code instead here: https://github.com/TeamHG-Memex/arachnado/pull/17/files#diff-cc14c41526bf9fb889d041d47c46cab4R41